File: python/08_039.py
def max_productivity_zone(capacities):

    stack = []
    max_area = 0
    n = len(capacities)
    
    for i in range(n):
        # Process stations with higher capacity than current
        while stack and capacities[stack[-1]] > capacities[i]:
            height = capacities[stack.pop()]
            width = i if not stack else i - stack[-1] - 1
            max_area = max(max_area, height * width)
        
        stack.append(i)
    
    # Process remaining stations in stack
    while stack:
        height = capacities[stack.pop()]
        width = n if not stack else n - stack[-1] - 1
        max_area = max(max_area, height * width)
    
    return max_area

File: java/08_039.java
import java.util.Stack;

public int maxProductivityZone(int[] capacities) {
    Stack<Integer> stack = new Stack<>();
    int maxArea = 0;
    int n = capacities.length;
    
    for (int i = 0; i < n; i++) {
        // Process stations with higher capacity than current
        while (!stack.isEmpty() && 
               capacities[stack.peek()] > capacities[i]) {
            int height = capacities[stack.pop()];
            int width = stack.isEmpty() ? i : i - stack.peek() - 1;
            maxArea = Math.max(maxArea, height * width);
        }
        
        stack.push(i);
    }
    
    // Process remaining stations in stack
    while (!stack.isEmpty()) {
        int height = capacities[stack.pop()];
        int width = stack.isEmpty() ? n : n - stack.peek() - 1;
        maxArea = Math.max(maxArea, height * width);
    }
    
    return maxArea;
}